vortex-serde/src/layouts/read/batch.rs

```diff
@@ -53,14 +116,221 @@ impl BatchReader {
+                        &filter_result,
+                        self.row_offset,
+                        filter_result.len(),
+                    )?;
+                    self.row_offset += b.len();
+                    Ok(Some(ReadResult::Selector(selector)))
+                }
+                ReadResult::Selector(_) => unreachable!("Can only produce batches"),
+            },
+        }
+    }
+
+    fn advance(&mut self, up_to_row: usize) -> VortexResult<Vec<Message>> {
+        self.reader.advance(up_to_row)
+    }
+}
+
+#[derive(Debug)]
+pub struct ColumnBatchFilter {
+    children: Vec<Box<dyn LayoutReader>>,
+    read_ranges: Vec<Option<RowSelector>>,
+    filters: Vec<Option<RowSelector>>,
+}
+
+impl ColumnBatchFilter {
+    pub fn new(children: Vec<Box<dyn LayoutReader>>) -> Self {
+        let read_ranges = vec![None; children.len()];
+        let filters = vec![None; children.len()];
+        Self {
+            children,
+            read_ranges,
+            filters,
+        }
+    }
+
+    pub fn next_range(&mut self) -> VortexResult<RangeResult> {
+        let mut messages = Vec::new();
+        for (i, child_selector) in self
+            .read_ranges
+            .iter_mut()
+            .enumerate()
+            .filter(|(_, a)| a.is_none())
+        {
+            match self.children[i].next_range()? {
+                RangeResult::ReadMore(m) => messages.extend(m),
+                RangeResult::Rows(s) => match s {
+                    None => return Ok(RangeResult::Rows(None)),
+                    Some(rs) => {
+                        if rs.is_empty() {
+                            return self.advance(rs.end()).map(RangeResult::ReadMore);
+                        }
+                        *child_selector = Some(rs);
+                    }
+                },
+            }
+        }
+
+        if messages.is_empty() {
+            let ranges = mem::replace(&mut self.read_ranges, vec![None; self.children.len()]);
+            let mut ranges_iter = ranges.iter().enumerate();
+            let mut final_range: Option<RowSelector> =
+                ranges_iter.next().and_then(|(_, ri)| ri.clone());
+            for (i, range) in ranges_iter {
+                let Some(column_range) = range else {
+                    vortex_bail!("Finished reading all columns but column {i} didn't produce range")
+                };
+                final_range = final_range.and_then(|fr| {
+                    let intersection = fr.intersect(column_range);
+                    if intersection.is_empty() {
+                        None
+                    } else {
+                        Some(intersection)
+                    }
+                })
+            }
+
+            if let Some(fr) = final_range.as_ref() {
+                self.read_ranges = ranges
+                    .into_iter()
+                    .map(|rs| rs.and_then(|r| r.advance(fr.end())))
+                    .collect();
+            }
```

Comment:
This logic can end up with small leftover batches

Reply:
I think we can put a floor on the chunk size

Reply:
I don't think you want to... I can put 10GB of binary data into a single cell. It gets put into a Vortex chunk with array length 1. I really don't want you to batch it!

Reply:
I had a different problem in mind, you observe that one array returned chunks of say 5 and the other returned 4. Right now we will return chunk of size 4 and then return chunk of size 1. However, we could store average of chunk sizes returned by each child and buffer until we have the average size. This is an optimization though we can do later.